### PR TITLE
Introduce CCA Realm provisioning plugin

### PR DESCRIPTION
**Pull request by yogeshbdeshpande**
This change introduces Realm Provisioning Plugin

### CODE REVIEW

**Comment by setrofim** [comment]
Shouldn't this be part of CCA scheme, rather than a separate scheme?

File: scheme/cca-realm-provisioning/corim_extractor.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	return refVals, nil
+}
+
+func makeRefValAttrs(cAttr *ClassAttributes,
+	iAttr *InstanceAttributes,
+	rAttr *RealmAttributes) (json.RawMessage, error) {
+
+	var attrs = map[string]interface{}{
+		"CCA_REALM.vendor":      cAttr.Vendor,
+		"CCA_REALM-class-id":    cAttr.UUID,
```
Comment (thomas-fossati) [inline scheme/cca-realm-provisioning/corim_extractor.go:68]: Is it the right choice?  Would "model" be a better identifier for what is essentially a software product?

File: scheme/cca-realm-provisioning/corim_extractor.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/veraison/corim/comid"
+	"github.com/veraison/services/handler"
+)
+
+type CorimExtractor struct{}
+
+func (o CorimExtractor) RefValExtractor(
```
Comment (setrofim) [inline scheme/cca-realm-provisioning/corim_extractor.go:15]: Rename to `ExtractRefVal`.  In general, methods should be verbs (unless constructors, factories, properties, etc). The name as-is suggests that this should return an extractor, rather than a slice of ref vals.

(note: I realise that this is meant to be implementing `common.IExtractor`, however there is no need for that interface -- this is only used within this code, and is never passed via the interface.)

File: scheme/cca-realm-provisioning/corim_extractor.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	data, err := json.Marshal(attrs)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal reference value attributes: %w", err)
+	}
+	return data, nil
+}
+
+func (o CorimExtractor) TaExtractor(
```
Comment (setrofim) [inline scheme/cca-realm-provisioning/corim_extractor.go:85]: Can just remove remove this? I don't think there is a need to implement ` common.IExtractor`?

File: scheme/cca-realm-provisioning/plugin/combined/main.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+package main
+
+import (
+	"github.com/veraison/services/handler"
+	"github.com/veraison/services/plugin"
+	scheme "github.com/veraison/services/scheme/cca-realm-provisioning"
+)
+
+func main() {
+	handler.RegisterEndorsementHandler(&scheme.EndorsementHandler{})
```
Comment (setrofim) [inline scheme/cca-realm-provisioning/plugin/combined/main.go:12]:
```suggestion
	handler.RegisterEndorsementHandler(&scheme.EndorsementHandler{})
	handler.RegisterStoreHandler(&scheme.StoreHandler{})
```

File: scheme/cca-realm-provisioning/store_handler.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, fmt.Errorf("unable to synthesize reference value abs-path: %w", err)
+	}
+
+	lookupKey := RefValLookupKey(SchemeName, tenantID, instID)
+	log.Debugf("Scheme %s Plugin TA Look Up Key= %s\n", SchemeName, lookupKey)
+	return []string{lookupKey}, nil
+
+}
+
+func RefValLookupKey(schemeName, tenantID, instID string) string {
```
Comment (setrofim) [inline scheme/cca-realm-provisioning/store_handler.go:55]: Can be changed to a local function (i. e. `refValLookupKey`)?

**Review by setrofim — CHANGES_REQUESTED** [review]

File: scheme/Makefile [hunk 6]
```diff
@@ -8,6 +8,7 @@ SUBDIR += psa-iot
 SUBDIR += tpm-enacttrust
 SUBDIR += parsec-tpm
 SUBDIR += parsec-cca
+SUBDIR += cca-realm-provisioning
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: @setrofim : As discussed only the `provisioning` will be a separate scheme, as it is coming from a different supply chain, the Verification will be taken care inside a combined CCA plugin, which verifies both Realm and Platform part together. Hence I chosen the name cca-realm-provisioning.

If I keep it just `cca-realm` then it will assume both Provisioning and Verifications are handled inside!

File: scheme/cca-realm-provisioning/realmattributes.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	return digests[0].AlgIDToString(), digests[0].HashValue, nil
+}
+
+func (o *RealmAttributes) extractRegisterIndexes(r *comid.IntegrityRegisters) error {
+	for k, val := range r.M {
+		a, d, err := o.extractRealmDigest(val)
+		if err != nil {
+			return errors.New("unable to extract digest for ")
```
Comment (thomas-fossati) [inline scheme/cca-realm-provisioning/realmattributes.go:43]: incomplete sentence

File: scheme/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: Mmm ok, that would mean that the `StoreHandler` implementation would essentially be split/duplicated across two schemes, as the key formats need to be aligned across both schemes (which breaks the conceptual integrity of a scheme -- schems ought to be independent of each other).

Are we sure this is the right way forward? Would it not be better to implement as single `cca` scheme that handles everything in a single set of plugins, and then use configuration and/or policy to control whether the platform or realm parts, or both, are handled by a particular deployment?

File: scheme/cca-realm-provisioning/realmattributes.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func (o RealmAttributes) isCompatibleAlgID(hashAlgID string) bool {
+	compatible := true
+	if o.HashAlgID != "" {
+		if hashAlgID != o.HashAlgID {
+			compatible = false
+		}
+	}
+	return compatible
```
Comment (thomas-fossati) [inline scheme/cca-realm-provisioning/realmattributes.go:80]: isn't it equivalent to:

```suggestion
	return o.HashAlgID == "" || hashAlgID == o.HashAlgID
```

**Review by thomas-fossati — COMMENTED** [review]
a couple of questions: 

* is it possible to provide an incomplete integrity register (e.g., if I don't care about REMs and only care about RIM)
* why is PV left out from the realm reference values?

File: scheme/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: Sure, I thought about StoreHandler part, while implementing the provisioning scheme and came to conclusion that the internal `logic` will be implemented in the `common` set of functions and both the schemes will invoke the same internal implementation thus essentially avoiding duplication! This is the best way of approaching as the CCA Scheme StoreHandler needs to fetch in sequence the RefValues for Platform and Realm in sequence, however this scheme will only do that latter!

File: scheme/cca-realm-provisioning/realmattributes.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	return nil
+}
+
+func (o *RealmAttributes) extractRealmDigest(digests comid.Digests) (algID string, hash []byte, err error) {
+
+	if err := digests.Valid(); err != nil {
+		return "", nil, fmt.Errorf("invalid digest: %v", err)
+	}
+	if len(digests) != 1 {
+		return "", nil, fmt.Errorf("invalid number %d for digest", len(digests))
```
Comment (thomas-fossati) [inline scheme/cca-realm-provisioning/realmattributes.go:33]:
```suggestion
		return "", nil, fmt.Errorf("expecting 1 digest, got %d", len(digests))
```

**Comment by thomas-fossati** [comment]
Follow up on this:

> * is it possible to provide an incomplete integrity register (e.g., if I don't care about REMs and only care about RIM)?

I think there should be a `RealmAttributes::Valid()` method that makes it clear what are the exact requirements in terms of what we are willing to ingest.

**Comment by yogeshbdeshpande** [comment]
> RealmAttributes

@thomas-fossati: I will look into this further! Basically my initial thought was REM slots will always be 4 in token, so we provide 4 to Supply Chain. They can use whatever they want to use 1,2,3, or all. The rest will be filled with `nil` values same as received in the token. When comparing, we compare the whole blob and should work, isn't it?

File: scheme/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: Conceptually, a "scheme" is supposed to encompass the entire attestation "lifecycle", i.e. provisioning and attestion.  It doesn't make sense to have a scheme with only provisioning, after all, the sole purpose of provisioning endorsements is so that they would later be used in verification.

> the Verification will be taken care inside a combined CCA plugin, which verifies both Realm and Platform part together.

If that is the case, then this should be part of that combined plugin as well (going back to my intial suggestion of renaming this). When implementing the verification part of it, we can then re-use the commmon functionality between this and the `cca-ssd-platform` scheme.

On the other hand, if the goal is to modify `cca-ssd-platfom` scheme to optinally handle the realm part as well, then (it should probably be renamed, and) this should be part of that too. 

Basically, it doesn't make sense for provisioning to exist in a scheme of its own, as  the only purpose of provisioning is to enable verification. So if realm verification on its own is not a thing, and we can only have plaform and platform+realm, then we should have two schemes: `cca-ssd-platform` and `cca` (or `cca-combined`).

**Comment by thomas-fossati** [comment]
> > RealmAttributes
> 
> @thomas-fossati: I will look into this further! Basically my initial thought was REM slots will always be 4 in token, so we provide 4 to Supply Chain. They can use whatever they want to use 1,2,3, or all. The rest will be filled with `nil` values same as received in the token. When comparing, we compare the whole blob and should work, isn't it?

Not really.  Suppose at run-time the workload app extends the REMs but the verifier is only in the "pure" TCB (i.e., platform + RIM).

If both RIM and REMs must be specified there is no way to tell the verifier to "ignore REMs".

File: scheme/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomas-fossati) [reply scheme/Makefile:11]: > On the other hand, if the goal is to modify cca-ssd-platfom scheme to optinally handle the realm part as well, then (it should probably be renamed, and) this should be part of that too.

this looks like the "natural" choice: if there is realm ref-values, then we also appraise the realm, otherwise we only appraise the platform.

[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: > > On the other hand, if the goal is to modify cca-ssd-platfom scheme to optinally handle the realm part as well, then (it should probably be renamed, and) this should be part of that too.
> 
> this looks like the "natural" choice: if there is realm ref-values, then we also appraise the realm, otherwise we only appraise the platform.


Yes, the renaming of cca_ssd_platform to cca_scheme is on my work agenda.

However the Realm Provisioning cannot be part of that scheme.

It has a different suply chain. All I can accept is to rename this scheme as cca-realm, if both you and @thomas-fossati are ok. However, we agreed we do not need to implement `EvidenceHandler` interface for this scheme, as this will be done as part of CCA Scheme Verification!

[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: > It has a different suply chain. 

Why does this necessiate a different scheme? The difference is encapsulated in the media type, no? So the provisioning plugin for the combined scheme can differentiate on, and handle, both media types, and therefore both supply chains.

[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: Realm has a different "lifecycle"

[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: Realm provisioning has no bearing and no connection to cca-platform lifecycle, it is completely independent! Fusing them together will be incorrect!

[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: OK, but again, why does necessiate a different scheme? You do not need separate schemes to handle CoRIMs of different formats, or coming from different sources -- this is a matter of media type, not scheme.

[the remark this answers is not in the document]
Comment (thomas-fossati) [reply scheme/Makefile:11]: As long as realm and platform can be supplied in independent CoRIMs from independent provisioners, using the same media type (i.e., scheme) for both has no drawbacks that I can foresee.

[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: > using the same media type (i.e., scheme) 
 
media type != scheme. A scheme can support multiple media types. But yes, depending on how things are implemented, it may not even be necessary to define multiple media types (if the one media type defines the union of the entries from the different suppliers, and all entries are optional). 

I can see the value of separate media types (i.e CoRIM profiles) targeting different entities within the supply chain; those can still be handled within one scheme though.

My point is that the sole purpose of provisining endorsements is so that the are used in verification. The verifier determines what needs to be provisioned for a successful verification. So it does not make sense to have a "scheme" that only defines provisioning. That "scheme" will always be tied to some other scheme (the one that defienes verification), and so is not an indepent attestation scheme by itself. Allowing this would just confuse the definition of "scheme" as a concept within the project, and will make deployment configuration more complicated and error prone (when enabling plugins for a scheme, the deployer would then need to check if there are any dependent schemes that also need to be enabled, and that their versions are compatible, etc).

[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: Media types defines the payload container. CoRIM for CCA Platform, is a different container than a CoRIM for Realm.

Using the same media type for both is incorrect, isn't it?

[the remark this answers is not in the document]
Comment (thomas-fossati) [reply scheme/Makefile:11]: > Using the same media type for both is incorrect, isn't it?

It's not incorrect, as long as the same CoRIM profile takes care of both platform and realm.

Instead, if we want to have separate profiles, we will have two media types.

[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: Yes, they should be separate media types/profiles.

[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/Makefile:11]: Yes, CCA Platform and CCA Realm are two separate profiles and hence we will need two media types!

[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/Makefile:11]: > It's not incorrect, as long as the same CoRIM profile takes care of both platform and realm.

They technically _could_ be handled with a single profile. However, I belive the arguement is that they _should_ be handled by separate profiles, as they're expected to be privided by different entitied within the supply chain, and that would simplify the profile that each entity needs to understand.

[the remark this answers is not in the document]
Comment (thomas-fossati) [reply scheme/Makefile:11]: there are clear practical advantages to separating the two.  I was just noting that it's not "incorrect" per see.

I think we are all in agreement about "two media types for one scheme"

File: scheme/cca-realm-provisioning/corim_extractor.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	if err := instAttrs.FromEnvironment(rv.Environment); err != nil {
+		return nil, fmt.Errorf("could not extract Realm instance attributes: %w", err)
+	}
+
+	// Each measurement is encoded in a measurement-map of a CoMID
+	// reference-triple-record.  Since a measurement-map can encode one or more
+	// measurements, a single reference-triple-record can carry as many
+	// measurements as needed. However for Realm Instance, only one measurement
+	// record is set, with both the "rim" & "rem" measurements carried in an
+	// integrity register
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/cca-realm-provisioning/corim_extractor.go:34]: I have corrected the comment now!

File: scheme/cca-realm-provisioning/corim_extractor.go (hunk 1 above)
Comment (yogeshbdeshpande) [reply scheme/cca-realm-provisioning/corim_extractor.go:68]: the `intent` here was to identify a Workload provider Identity which is launching `N` number of Realms, each with a Unique `RIM` Value.

File: scheme/cca-realm-provisioning/corim_extractor.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func makeRefValAttrs(cAttr *ClassAttributes,
+	iAttr *InstanceAttributes,
+	rAttr *RealmAttributes) (json.RawMessage, error) {
+
+	var attrs = map[string]interface{}{
+		"CCA_REALM.vendor":      cAttr.Vendor,
+		"CCA_REALM-class-id":    cAttr.UUID,
+		"CCA_REALM-instance-id": rAttr.Rim,
```
[the remark this answers is not in the document]
Comment (yogeshbdeshpande) [reply scheme/cca-realm-provisioning/corim_extractor.go:69]: Yes, it can be expressed this way and perhaps better, hence renamed it now!

File: scheme/cca-realm-provisioning/corim_extractor.go (hunk 2 above)
Comment (yogeshbdeshpande) [reply scheme/cca-realm-provisioning/corim_extractor.go:15]: IExtractor is the interface which is been implemented by CorimExtractor. To your comment: _however there is no need for that interface -- this is only used within this code_

is not true, `IExtractor `  interfaces are invoked from: UnsignedCorimDecoder.go
hence this is not a local function!

File: scheme/cca-realm-provisioning/corim_extractor.go (hunk 3 above)
Comment (yogeshbdeshpande) [reply scheme/cca-realm-provisioning/corim_extractor.go:85]: As IExtractor is `needed` this cannot be removed!

**Comment by yogeshbdeshpande** [comment]
Closing this PR as the changes `intended` for this Pull Request will be addressed using the PR #233
